--- ./CcspTr069Pa/source/CcspCwmpSession/ccsp_cwmp_sesso_mcoif.c
+++ ./CcspTr069Pa/source/CcspCwmpSession/ccsp_cwmp_sesso_mcoif.c	2015-04-22 14:22:31.186833009 -0600
@@ -4793,10 +4793,10 @@
         ULONG                       ulArraySize     = 1;
         int                         iStatus         = 0;
         SLAP_VARIABLE               slapVar;
-        char                        value[2]        = {'1', 0};
+        char                       value[] = "Router,Wifi,VoIP,Dect,MoCA";
 
         ParamValue.Name          = CCSP_NS_FACTORYRESET;
-        ParamValue.Tr069DataType = CCSP_CWMP_TR069_DATA_TYPE_Boolean;
+        ParamValue.Tr069DataType = CCSP_CWMP_TR069_DATA_TYPE_String;
         ParamValue.Value         = &slapVar;
 
         SlapInitVariable(&slapVar);
@@ -4924,7 +4924,8 @@
 CcspCwmpsoAsyncReboot()
 {
     AnscSleep(3*1000);
-    system("reboot");
+    system("/fss/gw/rdklogger/backupLogs.sh");
+    //system("reboot");
 }
 #endif
 
